### PR TITLE
Creating chips custom renderer, updating visual look.

### PR DESCRIPTION
**Pull request by gspencergoog**
This updates the visual look of the Chip class, so that it scales properly in the face of text scale (and `label` widget size) changes, and bases its height off of the `label` widget's height, constraining the other widgets it contains to be the same height.

To do this properly, I had to implement a custom render object that will measure the height of the label, so the guts of this class are basically rewritten.

In addition, to allow the circle avatar to increase in size when the chip does, I added minRadius and maxRadius arguments to it, and I updated its color handling to use the light/dark primary colors of the theme in a smart way instead of just using black and white.

Updated and added tests.
screenshot_20180316-124504

### CODE REVIEW

File: packages/flutter/lib/src/material/chip.dart [hunk 1]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+    _edgePadding,
+  );
+
+  final Map<_ChipSlot, RenderBox> slotToChild = <_ChipSlot, RenderBox>{};
+  final Map<RenderBox, _ChipSlot> childToSlot = <RenderBox, _ChipSlot>{};
+
+  TapGestureRecognizer _tap;
+
+  VoidCallback onDeleted;
+  Rect _deleteButtonRegion;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: Maybe this should be called _deleteButtonRect.

File: packages/flutter/lib/src/material/chip.dart [hunk 2]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+      _tapDownLocation = globalToLocal(details.globalPosition);
+    }
+  }
+
+  void _handleTap() {
+    if (_tapDownLocation == null) {
+      return;
+    }
+    if (deleteIcon != null && _deleteButtonRegion.contains(_tapDownLocation)) {
+      onDeleted();
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: Is there a corner case where onDeleted is null? Maybe the widget is rebuilt with onDeleted:null in between the press and tap events?

File: packages/flutter/lib/src/material/chip.dart [hunk 3]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+    return box == null ? 0.0 : box.getMinIntrinsicWidth(width);
+  }
+
+  static Size _boxSize(RenderBox box) => box == null ? Size.zero : box.size;
+
+  static BoxParentData _boxParentData(RenderBox box) => box.parentData;
+
+  @override
+  double computeMinIntrinsicWidth(double height) {
+    // The overall padding isn't affected by missing avatar or delete icon
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: Thank you for this comment!

File: packages/flutter/lib/src/material/chip.dart [hunk 4]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+
+  @override
+  void performLayout() {
+    double iconHeight = _chipHeight;
+    if (label != null) {
+      label.layout(constraints.loosen(), parentUsesSize: true);
+      // Now that we know the height, we can determine how much to shrink the
+      // constraints by for the "real" layout. Ignored if the constraints are
+      // infinite.
+      iconHeight = math.max(iconHeight, _boxSize(label).height);
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: This is confusing in light of the comment on line 588 about the widget being sized to the height of the label only. If that were true, shouldn't this be `math.min(...)`?

File: packages/flutter/lib/src/material/chip.dart [hunk 5]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+      if (constraints.maxWidth.isFinite) {
+        final double allPadding = _iconPadding.horizontal * 2.0 + theme.labelPadding.horizontal;
+        final double iconSizes = (avatar != null ? iconHeight - _iconPadding.horizontal : 0.0)
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: Shouldn't this be `iconHeight - _iconPadding.vertical`? Likewise for the next line.

File: packages/flutter/lib/src/material/chip.dart [hunk 6]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+            + (deleteIcon != null ? iconHeight - _iconPadding.horizontal : 0.0);
+        label.layout(
+          constraints.loosen().copyWith(
+                maxWidth: math.max(0.0, constraints.maxWidth - iconSizes - allPadding),
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: Still a bit confusing here. Started with `iconHeight`, now we're using it to set a widget constraint.

Maybe this would be clearer if the iconHeight was called iconSize (because it occupies a iconSize x iconSize rect)?

File: packages/flutter/lib/src/material/chip.dart [hunk 7]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+  );
+
+  final Map<_ChipSlot, RenderBox> slotToChild = <_ChipSlot, RenderBox>{};
+  final Map<RenderBox, _ChipSlot> childToSlot = <RenderBox, _ChipSlot>{};
+
+  TapGestureRecognizer _tap;
+
+  VoidCallback onDeleted;
+  Rect _deleteButtonRegion;
+  Rect _actionRegion;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: Likewise, actionRect

File: packages/flutter/lib/src/material/chip.dart [hunk 8]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+          context.canvas.drawRect(
+            _actionRegion.shift(offset),
+            outlinePaint..color = const Color(0xff008000),
+          );
+          return true;
+        }());
+
+    doPaint(container);
+    doPaint(avatar);
+    if (deleteIcon != null) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip.dart]: Don't need this test because doPaint()

File: packages/flutter/lib/src/material/circle_avatar.dart [hunk 9]
```diff
@@ -67,13 +69,19 @@ class CircleAvatar extends StatelessWidget {
   /// The color with which to fill the circle. Changing the background
   /// color will cause the avatar to animate to the new color.
   ///
-  /// If a background color is not specified, the theme's primary color is used.
+  /// If a background color is not specified, the theme's light primary color
+  /// is used with dark foreground colors, and the dark primary color with light
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/circle_avatar.dart]: Although it's a bit clunky to write, it would help to write [ThemeData.primaryColorDark] instead of "dark primary color" etc.

File: packages/flutter/lib/src/material/circle_avatar.dart [hunk 10]
```diff
@@ -67,13 +69,19 @@ class CircleAvatar extends StatelessWidget {
...
   final Color backgroundColor;
 
   /// The default text color for text in the circle.
   ///
-  /// Falls back to white if a background color is specified, or the primary
-  /// text theme color otherwise.
+  /// Defaults to the primary text theme color if no [backgroundColor] is
+  /// specified.
+  ///
+  /// If a [backgroundColor] is specified, this falls back to the light primary
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/circle_avatar.dart]: If _no_ [backgroundColor] is specified ... then [foregroundColor] defaults to ...

File: packages/flutter/lib/src/material/circle_avatar.dart [hunk 11]
```diff
@@ -85,48 +93,113 @@ class CircleAvatar extends StatelessWidget {
   /// The size of the avatar. Changing the radius will cause the
   /// avatar to animate to the new size.
   ///
+  /// If radius is specified, then [minRadius] and [maxRadius] may not be
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/circle_avatar.dart]: may not => must not

File: packages/flutter/lib/src/material/circle_avatar.dart [hunk 12]
```diff
@@ -85,48 +93,113 @@ class CircleAvatar extends StatelessWidget {
...
+  ///
+  /// Changing the maxRadius will cause the avatar to animate to the new size,
+  /// if constraints allow.
+  ///
+  /// If maxRadius is specified, then [radius] may not also be specified.
+  ///
+  /// Defaults to [double.infinity].
+  final double maxRadius;
+
+  static const double _defaultRadius = 20.0;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/circle_avatar.dart]: _kConstants

File: packages/flutter/lib/src/material/circle_avatar.dart [hunk 13]
```diff
@@ -85,48 +93,113 @@ class CircleAvatar extends StatelessWidget {
...
+  static const double _defaultMaxRadius = double.infinity;
+
+  double get _minDiameter {
+    double minDiameter = _defaultRadius * 2.0;
+    if (radius != null) {
+      minDiameter = radius * 2.0;
+    } else if (minRadius != null || maxRadius != null) {
+      minDiameter = 2.0 * (minRadius ?? _defaultMinRadius);
+    }
+    return minDiameter;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/circle_avatar.dart]: Perhaps this could be boiled down a little. It looks like:
```
 double get _minDiameter => 2.0 *  (_radius ?? _minRadius ?? _defaultMinRadius);
```

File: packages/flutter/lib/src/material/circle_avatar.dart [hunk 14]
```diff
@@ -85,48 +93,113 @@ class CircleAvatar extends StatelessWidget {
...
+  }
+
+  double get _maxDiameter {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/circle_avatar.dart]: Perhaps this could be boiled down a little. It looks like:
```
 double get _maxDiameter => 2.0 *  (_radius ?? _maxRadius ?? _defaultMaxRadius);
```

**Review by HansMuller — COMMENTED** [review]
This looks good. All comments are just nits.

File: packages/flutter/lib/src/material/chip.dart [hunk 15]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+  final VoidCallback onDeleted;
+
+  @override
+  _RenderChipElement createElement() => new _RenderChipElement(this);
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderChip renderObject) {
+    renderObject
+      ..theme = theme
+      ..textDirection = Directionality.of(context);
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Oooh!  Thanks for that!

File: packages/flutter/lib/src/material/chip.dart [hunk 16]
```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
...
+        _textDirection = textDirection {
+    _tap = new TapGestureRecognizer(debugOwner: this)
+      ..onTapDown = _handleTapDown
+      ..onTap = _handleTap;
+  }
+
+  // Set this to true to have outlines of the tap targets drawn over
+  // the chip.  This should never be checked in while set to 'true'.
+  static const bool _debugShowTapTargetOutlines = false;
+  static const EdgeInsetsDirectional _iconPadding = const EdgeInsetsDirectional.fromSTEB(
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Done.

File: packages/flutter/lib/src/material/chip.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Done.

File: packages/flutter/lib/src/material/chip.dart (hunk 7 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Done.

File: packages/flutter/lib/src/material/chip.dart (hunk 2 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Good catch. Yeah, unlikely but possible.

File: packages/flutter/lib/src/material/chip.dart (hunk 4 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Hmm, actually, it's the computeMinIntrinsicHeight that is wrong.  It should have a max in it.  Fixed (I think).

File: packages/flutter/lib/src/material/chip.dart (hunk 5 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Yep (I already fixed that in my next PR, doing it here too).

File: packages/flutter/lib/src/material/chip.dart (hunk 6 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Well, yeah, but I used that below.  I switched to using the existing iconSize below, and adding the correct padding back in.

File: packages/flutter/lib/src/material/chip.dart (hunk 8 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Good point.

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 9 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/circle_avatar.dart]: OK.

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 10 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/circle_avatar.dart]: Actually, no, it only defaulted to those if backgroundColor is not null, but I'm rethinking that.  The old code forced black and white as colors, and I don't want to do that, but the way I had it, it would ignore the foreground color you set if you also set the background color, and that's not OK.  I think I fixed it so that if you set both, you get what you set, but if you set one, then we pick a contrasting one for you for the other one.

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 12 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/circle_avatar.dart]: Not for member constants, only globals. https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#begin-global-constant-names-with-prefix-k

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 13 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/circle_avatar.dart]: Yes, thanks, that is simpler, but if they are ALL null, then min and max should both be set to the _defaultRadius, which your version ignores.

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 11 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/circle_avatar.dart]: Done.

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 13 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/circle_avatar.dart]: I see. So shouldn't this just be (and you don't need _defaultMinRadius, _defaultMaxRadius)
```
double get _minDiameter => 2.0 *  (_radius ?? _minRadius ?? _defaultRadius);
```

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 10 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/circle_avatar.dart]: I see, so the foreground color is automatically computed based on the background color. NICE

File: packages/flutter/lib/src/material/circle_avatar.dart [hunk 17]
```diff
@@ -85,48 +93,113 @@ class CircleAvatar extends StatelessWidget {
...
-          // Need to reset the textScaleFactor here so that the
-          // text doesn't escape the avatar when the textScaleFactor is large.
-          data: MediaQuery.of(context).copyWith(textScaleFactor: 1.0),
-          child: new DefaultTextStyle(
-            style: textStyle.copyWith(color: foregroundColor),
-            child: child,
-          ),
-        )
-      ) : null,
+      child: child != null
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/circle_avatar.dart]: I think this kind of thing is slightly easier to read if you put the child == null so return null case first. Like
```
child: child == null
  ? null,
  : new Center(
      ...
    )
```

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 17 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/circle_avatar.dart]: Yes, I see what you mean.  Fixed.

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 13 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/circle_avatar.dart]: No, because if you specify just the `maxRadius` and `minRadius` is null, then the `minRadius` should be zero, not the default radius.  At least that's how I wanted it to work: If you specify nothing, then you get a default equivalent to `radius: _defaultRadius`.  If you specify a `radius`, then you get exactly that radius (as before), and if you specify only one of max or min, then the other one is at the limit.

File: packages/flutter/lib/src/material/chip.dart [hunk 18]
```diff
@@ -40,48 +53,46 @@ import 'tooltip.dart';
 ///  * [CircleAvatar], which shows images or initials of people.
 ///  * <https://material.google.com/components/chips.html>
 class Chip extends StatelessWidget {
-  /// Creates a material design chip.
+  /// Creates a material design chip
```
Comment (Hixie) [inline packages/flutter/lib/src/material/chip.dart]: missing a period...

File: packages/flutter/lib/src/material/chip.dart [hunk 19]
```diff
@@ -101,83 +113,642 @@ class Chip extends StatelessWidget {
...
-  ///
-  /// This has no effect when [onDelete] is null since no delete icon will be
-  /// shown.
+  /// Color for delete icon. The default is based on the ambient [IconTheme].
   final Color deleteIconColor;
 
   /// Message to be used for the chip delete button's tooltip.
-  ///
-  /// This has no effect when [onDelete] is null since no delete icon will be
-  /// shown.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/chip.dart]: why did you remove these "has no effect" paragraphs?

File: packages/flutter/lib/src/material/tooltip.dart [hunk 20]
```diff
@@ -170,9 +170,9 @@ class _TooltipState extends State<Tooltip> with SingleTickerProviderStateMixin {
 
   @override
   void deactivate() {
+    super.deactivate();
     if (_entry != null)
       _controller.reverse();
-    super.deactivate();
```
Comment (Hixie) [inline packages/flutter/lib/src/material/tooltip.dart]: can you elaborate on this change? It's unusual to do work after a deactivate, since by then you're not active and shouldn't do work...

File: packages/flutter/lib/src/material/chip.dart [hunk 21]
```diff
@@ -101,83 +113,642 @@ class Chip extends StatelessWidget {
...
+        _theme = theme,
+        _textDirection = textDirection {
+    _tap = new TapGestureRecognizer(debugOwner: this)
+      ..onTapDown = _handleTapDown
+      ..onTap = _handleTap;
+  }
+
+  // Set this to true to have outlines of the tap targets drawn over
+  // the chip.  This should never be checked in while set to 'true'.
+  static const bool _debugShowTapTargetOutlines = false;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/chip.dart]: use debugPaintSizeEnabled and the debugPaint method on render objects for this kind of logic.

File: packages/flutter/lib/src/material/chip.dart (hunk 19 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/chip.dart]: Because Hans asked me to... :-)

Comment (HansMuller) [reply packages/flutter/lib/src/material/chip.dart]: Hans felt that we were overstating the obvious and compounding the problem with verbiage.
